Fix various filehandle leaks

cachedout opened the pull request:
Reverts cf4f926 and modifies 63b1959 to use a weakref so that the bound method can be garbage collected.

cc: @driskell and @jacksontj

Commit:
Use weakref for bound function to avoid filehandle leaks

Commit:
Revert c96d4f7

Commit:
Lint and merge conflict

s0undt3ch commented:
Nice! I was messing on a lot more stuff in my attempt. Clean!

jacksontj commented:
This will help, but IIRC the main problem is that we are using singletons and the tests don't ever clear them (since it basically creates LOTS of IOLoops and never seems to close them). If this works worst case its a performance degradation (which we already have one which is worse) so it should be fine ;)

s0undt3ch commented:
The tests failing to clean up, and not salt _per se_ is something that should be addressed in the reet suite. Ie, force GC, force del, etc...

s0undt3ch commented:
Test suite

s0undt3ch commented:
Ie, we need the regular salt routines cleaning themselves and not leaking. The tests suite also not cleaning up might be a sign that this is not the complete fix.

Commit:
Switch from weakref to static methods

This still prevents leaks and avoids a situation where the proxy is getting dereferenced too early.

cachedout commented:
I had to switch this to use static methods. The weakref approach was dereferencing too early.
